Add teardown method to BaseProfiler.

camruta opened the pull request:
## What does this PR do?

This PR adds a teardown method to BaseProfiler class. With the method, the different instances can handle arbitrary post-profiling steps outside of __del__, such as closing output file how they want to, or do the log stitching.

This was brought up in PR 6067 review here:
https://github.com/PyTorchLightning/pytorch-lightning/pull/6067/files/928e680a24a0483bff3ebef100c531adc09acce3?file-filters%5B%5D=.py#r583854357

ananthsub reviewed (commented):
thanks for working on this @camruta ! A few things we should add:
- Where in the trainer do you want to call `teardown` 
- Unit tests to exercise that teardown works as expected for each of the provided profilers

camruta commented:
> thanks for working on this @camruta ! A few things we should add:
> 
> * Where in the trainer do you want to call `teardown`

As suggested by you in the offline discussion, we will call this method in trainer before calling the call_teardown_hook. I'll update the code.

> * Unit tests to exercise that teardown works as expected for each of the provided profilers

Adding unit tests.

tchaton reviewed (commented):
LGTM ! Please, add a test to make sure the function is being called.

pep8speaks commented:
Hello @camruta! Thanks for updating this PR. 







* In the file [`tests/trainer/test_trainer.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/af608f05f6ea6f8f76c72d9679ad9fedd62aee3a/tests/trainer/test_trainer.py):

> [Line 1606:1](https://github.com/PyTorchLightning/pytorch-lightning/blob/af608f05f6ea6f8f76c72d9679ad9fedd62aee3a/tests/trainer/test_trainer.py#L1606): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

camruta commented:
> LGTM ! Please, add a test to make sure the function is being called.

Thanks for the review! Added the unite test.

camruta commented:
> thanks for working on this @camruta ! A few things we should add:
> 
> * Where in the trainer do you want to call `teardown`
Added profiler.teardown call in trainer.py.
> * Unit tests to exercise that teardown works as expected for each of the provided profilers
Added unite test.

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    This test checks if profiler teardown method is called when
+    trainer is exiting.
+    """
+    profilerSimple = SimpleProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
+    profilerAdvanced = AdvancedProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
+    profilerPytorch = PyTorchProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"), local_rank=0)
+
```
```suggestion
    simple_profiler = SimpleProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
    advanced_profiler = AdvancedProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
    pytorch_profiler = PyTorchProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"), local_rank=0)

```

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profilerSimple,
```
```suggestion
        profiler=simple_profiler,
```

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    )
+    trainer.fit(model)
+
+    assert profilerSimple.output_file.closed
```
```suggestion
    assert simple_profiler.output_file.closed
```

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    )
+    trainer.fit(model)
+
+    assert profilerAdvanced.output_file.closed
```
```suggestion
    assert advanced_profiler.output_file.closed
```

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    )
+    trainer.fit(model)
+
+    assert profilerPytorch.output_file.closed
```
```suggestion
    assert pytorch_profiler.output_file.closed
```

ananthsub commented on tests/trainer/test_trainer.py:
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
     trainer.fit(model)
 
 
+def test_profiler_teardown(tmpdir):
```
you could add a dummy profiler which sets profiler state in `teardown` which you can use to verify in the unit test that `teardown` is called

do we have a separate `test_profilers.py` ? i think it'd be useful to split out from the trainer. what do you think @tchaton ?

carmocca replied:
> do we have a separate test_profilers.py ? i think it'd be useful to split out from the trainer

Yes! To `tests/test_profiler.py`

I have actually moved them in my (very) WIP PR to refactor profilers

carmocca replied:
Moved in https://github.com/PyTorchLightning/pytorch-lightning/commit/51c9260fad5b1ed3b4e41a9ebf460bf4c609fe2a

ananthsub reviewed (commented):
mostly style nits and a comment on unit testing profilers

carmocca reviewed (approved):
Merged master and applied some fixes! LGTM

ananthsub commented on CHANGELOG.md:
```diff
@@ -37,6 +39,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added arg to `self.log` that enables users to give custom names when dealing with multiple dataloaders ([#6274](https://github.com/PyTorchLightning/pytorch-lightning/pull/6274))
 
 
+- Added `teardown` method to `BaseProfiler` to enable subclasses defe post-profiling steps outside of `__del__` ([#6370](https://github.com/PyTorchLightning/pytorch-lightning/pull/6370))
```
```suggestion
- Added `teardown` method to `BaseProfiler` to enable subclasses to do post-profiling steps outside of `__del__` ([#6370](https://github.com/PyTorchLightning/pytorch-lightning/pull/6370))
```

tchaton reviewed (approved):
LGTM !
